kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -201,13 +200,50 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches(
    * use the P and/or M from the hint.
    */
   protected Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
-      Engine engine, Optional<SnapshotHint> snapshotHint, long snapshotVersion) {
+      Engine engine,
+      LogSegment logSegment,
+      Optional<SnapshotHint> snapshotHint,
+      long snapshotVersion) {
 
-    // Exit early if the hint already has the info we need
+    // Exit early if the hint already has the info we need.
     if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
       return new Tuple2<>(snapshotHint.get().getProtocol(), snapshotHint.get().getMetadata());
     }
 
+    // Snapshot hit is not use-able in this case for determine the lower bound.
+    if (snapshotHint.isPresent() && snapshotHint.get().getVersion() > snapshotVersion) {
+      snapshotHint = Optional.empty();
+    }
+
+    // Finds the exclusive lower bound for CRC search.
+    // If the snapshot hint or checkpoint older than required version is present, we can use them as
+    // the lower bound for the CRC search.
+    long crcSearchLowerBound =
+        Math.min(
+            snapshotVersion,
+            Collections.max(
+                Arrays.asList(
+                    snapshotHint.map(SnapshotHint::getVersion).orElse(0L) + 1,
+                    logSegment.checkpointVersionOpt.orElse(0L),
+                    // Only find the CRC within 100 versions.
+                    Math.max(0, snapshotVersion - 100))));
```

Comment:
I find this confusing to read.

Didn't we already eliminate the `snapshotHint` if it's greater than `snapshotVersion`? So I think the `min` here might be unnecessary?

Also I think we don't want to use `checkpointVersionOpt` but rather the version from `checkpoints`? which will be guaranteed to be less than or equal to `snapshotVersion` -- so also no need for this min operation.

I think all that needs to happen here is
```
Math.max(
   checkpointVersionFromLogSegment,
   snapshotVersion - 100,
   snapshotHint.orElse(0) + 1,
   0
)
```

Reply:
corner case is snapshotVersion == 0 :-( so using min is to ensure snapshotVersion>= lowerbound for corner case like this

Reply:
Good catch Allison! The `checkpointVersionOpt` was my idea, and as Allison pointed out this is indeed incorrect. The version from the LogSegments.checkpoints is correct

+1 on we should make this easier to understand

Reply:
Updated, removed min() and use LogSegments.checkpoints. Thanks both.

Reply:
Update: We do actually want to use `checkpointVersionOpt`. It should have the save version as `checkpoints`